File: targets/targets_test.go
package targets_test

import (
	"strings"
	"testing"

	"github.com/retbrown/archeryutils/targets"
)

func mustTarget(t *testing.T, sys targets.ScoringSystem, diam, dist targets.Quantity, indoor bool) *targets.Target {
	t.Helper()
	tgt, err := targets.NewTarget(sys, diam, dist, indoor)
	if err != nil {
		t.Fatalf("NewTarget(%q,...) error: %v", sys, err)
	}
	return tgt
}

// ---- TestTarget ----

func TestTargetString(t *testing.T) {
	tgt := mustTarget(t, targets.TenZone, targets.CM(80), targets.Metres(30), false)
	want := "Target('10_zone', (80, 'cm'), (30, 'metre'), indoor=false)"
	if got := tgt.String(); got != want {
		t.Errorf("String() = %q, want %q", got, want)
	}
}

func TestTargetStringNativeUnits(t *testing.T) {
	tgt := mustTarget(t, targets.Worcester, targets.Inches(16), targets.Yards(20), true)
	want := "Target('Worcester', (16, 'inch'), (20, 'yard'), indoor=true)"
	if got := tgt.String(); got != want {
		t.Errorf("String() = %q, want %q", got, want)
	}
}

func TestTargetEquality(t *testing.T) {
	base := mustTarget(t, targets.TenZone, targets.CM(40), targets.Metres(20), true)

	tests := []struct {
		name   string
		args   func() *targets.Target
		want   bool
	}{
		{"duplicate", func() *targets.Target {
			return mustTarget(t, targets.TenZone, targets.CM(40), targets.Metres(20), true)
		}, true},
		{"units-free-cm-metre", func() *targets.Target {
			// no explicit unit → CM(40)/Metres(20) equivalent
			return mustTarget(t, targets.TenZone, targets.CM(40), targets.Metres(20), true)
		}, true},
		{"different-loc", func() *targets.Target {
			return mustTarget(t, targets.TenZone, targets.CM(40), targets.Metres(20), false)
		}, false},
		{"different-scoring", func() *targets.Target {
			return mustTarget(t, targets.FiveZone, targets.CM(40), targets.Metres(20), true)
		}, false},
		{"different-dist", func() *targets.Target {
			return mustTarget(t, targets.TenZone, targets.CM(40), targets.Metres(19.9), true)
		}, false},
		{"different-diam", func() *targets.Target {
			return mustTarget(t, targets.TenZone, targets.CM(40.1), targets.Metres(20), true)
		}, false},
		{"different-dist-unit", func() *targets.Target {
			return mustTarget(t, targets.TenZone, targets.CM(40), targets.Yards(20), true)
		}, false},
		{"different-diam-unit", func() *targets.Target {
			return mustTarget(t, targets.TenZone, targets.Inches(40), targets.Metres(20), true)
		}, false},
	}

	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			if got := base.Equal(tc.args()); got != tc.want {
				t.Errorf("Equal() = %v, want %v", got, tc.want)
			}
		})
	}
}

func TestTargetInvalidSystem(t *testing.T) {
	_, err := targets.NewTarget("InvalidScoringSystem", targets.CM(122), targets.Metres(50), false)
	if err == nil {
		t.Fatal("expected error for invalid system")
	}
	if !strings.Contains(err.Error(), "Invalid Target Face Type specified") {
		t.Errorf("error %q does not mention 'Invalid Target Face Type specified'", err.Error())
	}
}

func TestTargetInvalidDistanceUnit(t *testing.T) {
	_, err := targets.NewTarget(targets.FiveZone, targets.CM(122), targets.Quantity{50, "InvalidUnit"}, false)
	if err == nil {
		t.Fatal("expected error for invalid distance unit")
	}
}

func TestTargetDefaultDistanceUnit(t *testing.T) {
	tgt := mustTarget(t, targets.FiveZone, targets.CM(122), targets.Metres(50), false)
	if tgt.NativeDistance().Units != "metre" {
		t.Errorf("NativeDistance().Units = %q, want metre", tgt.NativeDistance().Units)
	}
}

func TestTargetYardToMetreConversion(t *testing.T) {
	tgt := mustTarget(t, targets.FiveZone, targets.CM(122), targets.Yards(50), false)
	if tgt.NativeDistance().Units != "yard" {
		t.Errorf("NativeDistance().Units = %q, want yard", tgt.NativeDistance().Units)
	}
	if !approxEqual(tgt.Distance(), 50.0*0.9144, 1e-9) {
		t.Errorf("Distance() = %v, want %v", tgt.Distance(), 50.0*0.9144)
	}
}

func TestTargetDefaultDiamUnit(t *testing.T) {
	tgt := mustTarget(t, targets.TenZoneFiveRingCmpd, targets.CM(80), targets.Metres(50), false)
	if !approxEqual(tgt.Diameter(), 80*0.01, 1e-9) {
		t.Errorf("Diameter() = %v, want %v", tgt.Diameter(), 80*0.01)
	}
}

func TestTargetDiamMetresNotConverted(t *testing.T) {
	tgt := mustTarget(t, targets.BeiterHitMiss, targets.Metres(0.04), targets.Metres(18), false)
	if tgt.Diameter() != 0.04 {
		t.Errorf("Diameter() = %v, want 0.04", tgt.Diameter())
	}
}

func TestTargetInchDiameter(t *testing.T) {
	tgt := mustTarget(t, targets.Worcester, targets.Inches(16), targets.Yards(20), true)
	if tgt.NativeDiameter().Units != "inch" {
		t.Errorf("NativeDiameter().Units = %q, want inch", tgt.NativeDiameter().Units)
	}
	if !approxEqual(tgt.Diameter(), 16*0.0254, 1e-9) {
		t.Errorf("Diameter() = %v, want %v", tgt.Diameter(), 16*0.0254)
	}
}

func TestTargetUnitsCoercedToDefinitive(t *testing.T) {
	imperial := mustTarget(t, targets.Worcester,
		targets.Quantity{16, "Inches"}, targets.Quantity{20, "Yards"}, true)
	metric := mustTarget(t, targets.TenZone,
		targets.Quantity{80, "Centimetres"}, targets.Quantity{30, "Metres"}, false)

	if imperial.NativeDistance().Units != "yard" {
		t.Errorf("imperial NativeDistance().Units = %q, want yard", imperial.NativeDistance().Units)
	}
	if imperial.NativeDiameter().Units != "inch" {
		t.Errorf("imperial NativeDiameter().Units = %q, want inch", imperial.NativeDiameter().Units)
	}
	if metric.NativeDistance().Units != "metre" {
		t.Errorf("metric NativeDistance().Units = %q, want metre", metric.NativeDistance().Units)
	}
	if metric.NativeDiameter().Units != "cm" {
		t.Errorf("metric NativeDiameter().Units = %q, want cm", metric.NativeDiameter().Units)
	}
}

func TestTargetDefaultLocation(t *testing.T) {
	tgt := mustTarget(t, targets.FiveZone, targets.CM(122), targets.Metres(50), false)
	if tgt.Indoor() {
		t.Error("Indoor() = true, want false")
	}
}

func TestTargetMaxScore(t *testing.T) {
	tests := []struct {
		system targets.ScoringSystem
		want   float64
	}{
		{targets.FiveZone, 9},
		{targets.TenZone, 10},
		{targets.TenZoneCompound, 10},
		{targets.TenZoneSixRing, 10},
		{targets.TenZoneFiveRing, 10},
		{targets.TenZoneFiveRingCmpd, 10},
		{targets.ElevenZone, 11},
		{targets.ElevenZoneSixRing, 11},
		{targets.ElevenZoneFiveRing, 11},
		{targets.WAField, 6},
		{targets.IFAAField, 5},
		{targets.IFAAFieldExpert, 5},
		{targets.AANationalField, 5},
		{targets.Worcester, 5},
		{targets.Worcester2Ring, 5},
		{targets.BeiterHitMiss, 1},
	}
	for _, tc := range tests {
		t.Run(string(tc.system), func(t *testing.T) {
			tgt := mustTarget(t, tc.system, targets.CM(122), targets.Metres(50), false)
			if got := tgt.MaxScore(); got != tc.want {
				t.Errorf("MaxScore() = %v, want %v", got, tc.want)
			}
		})
	}
}

func TestTargetMinScore(t *testing.T) {
	tests := []struct {
		system targets.ScoringSystem
		want   float64
	}{
		{targets.FiveZone, 1},
		{targets.TenZone, 1},
		{targets.TenZoneCompound, 1},
		{targets.TenZoneSixRing, 5},
		{targets.TenZoneFiveRing, 6},
		{targets.TenZoneFiveRingCmpd, 6},
		{targets.ElevenZone, 1},
		{targets.ElevenZoneSixRing, 5},
		{targets.ElevenZoneFiveRing, 6},
		{targets.WAField, 1},
		{targets.IFAAField, 3},
		{targets.IFAAFieldExpert, 1},
		{targets.AANationalField, 1},
		{targets.Worcester, 1},
		{targets.Worcester2Ring, 4},
		{targets.BeiterHitMiss, 1},
	}
	for _, tc := range tests {
		t.Run(string(tc.system), func(t *testing.T) {
			tgt := mustTarget(t, tc.system, targets.CM(122), targets.Metres(50), false)
			if got := tgt.MinScore(); got != tc.want {
				t.Errorf("MinScore() = %v, want %v", got, tc.want)
			}
		})
	}
}

func TestTargetFaceSpec(t *testing.T) {
	tests := []struct {
		system targets.ScoringSystem
		diam   targets.Quantity
		want   targets.FaceSpec
	}{
		{targets.FiveZone, targets.CM(122), targets.FaceSpec{
			0.244: 9, 0.488: 7, 0.732: 5, 0.976: 3, 1.22: 1,
		}},
		{targets.TenZone, targets.CM(80), targets.FaceSpec{
			0.08: 10, 0.16: 9, 0.24: 8, 0.32: 7, 0.40: 6,
			0.48: 5, 0.56: 4, 0.64: 3, 0.72: 2, 0.80: 1,
		}},
		{targets.ElevenZone, targets.CM(40), targets.FaceSpec{
			0.02: 11, 0.04: 10, 0.08: 9, 0.12: 8, 0.16: 7,
			0.20: 6, 0.24: 5, 0.28: 4, 0.32: 3, 0.36: 2, 0.40: 1,
		}},
		{targets.ElevenZoneSixRing, targets.CM(40), targets.FaceSpec{
			0.02: 11, 0.04: 10, 0.08: 9, 0.12: 8, 0.16: 7, 0.20: 6, 0.24: 5,
		}},
		{targets.WAField, targets.CM(80), targets.FaceSpec{
			0.08: 6, 0.16: 5, 0.32: 4, 0.48: 3, 0.64: 2, 0.80: 1,
		}},
		{targets.IFAAField, targets.CM(50), targets.FaceSpec{
			0.10: 5, 0.30: 4, 0.50: 3,
		}},
		{targets.AANationalField, targets.CM(20), targets.FaceSpec{
			0.04: 5, 0.08: 4, 0.12: 3, 0.16: 2, 0.20: 1,
		}},
		{targets.BeiterHitMiss, targets.CM(6), targets.FaceSpec{0.06: 1}},
		{targets.Worcester, targets.Inches(16), targets.FaceSpec{
			0.08128: 5, 0.16256: 4, 0.24384: 3, 0.32512: 2, 0.40640: 1,
		}},
		{targets.TenZoneSixRing, targets.CM(80), targets.FaceSpec{
			0.08: 10, 0.16: 9, 0.24: 8, 0.32: 7, 0.40: 6, 0.48: 5,
		}},
		{targets.TenZoneFiveRingCmpd, targets.CM(40), targets.FaceSpec{
			0.02: 10, 0.08: 9, 0.12: 8, 0.16: 7, 0.20: 6,
		}},
	}

	for _, tc := range tests {
		t.Run(string(tc.system), func(t *testing.T) {
			tgt := mustTarget(t, tc.system, tc.diam, targets.Metres(30), false)
			got, err := tgt.FaceSpec()
			if err != nil {
				t.Fatalf("FaceSpec() error: %v", err)
			}
			if !faceSpecEqual(got, tc.want, 1e-5) {
				t.Errorf("FaceSpec() = %v, want %v", got, tc.want)
			}
		})
	}
}

func TestTargetFaceSpecWrongConstructor(t *testing.T) {
	tgt := mustTarget(t, targets.Custom, targets.CM(122), targets.Metres(50), false)
	_, err := tgt.FaceSpec()
	if err == nil {
		t.Fatal("expected error accessing face spec on Custom target with no spec")
	}
	if !strings.Contains(err.Error(), "Trying to generate face spec for custom target") {
		t.Errorf("unexpected error: %v", err)
	}
}

func TestGenFaceSpecUnsupported(t *testing.T) {
	_, err := targets.GenFaceSpec("Dartchery", 1.0)
	if err == nil {
		t.Fatal("expected error for unsupported scoring system")
	}
}

// ---- TestCustomScoringTarget ----

func TestCustomTargetConstructor(t *testing.T) {
	spec := targets.FaceSpec{0.1: 3, 0.5: 1}
	tgt, err := targets.FromFaceSpec(spec, targets.CM(80), targets.Yards(50), false)
	if err != nil {
		t.Fatalf("FromFaceSpec error: %v", err)
	}
	if !approxEqual(tgt.Distance(), 50.0*0.9144, 1e-9) {
		t.Errorf("Distance() = %v, want %v", tgt.Distance(), 50.0*0.9144)
	}
	if tgt.Diameter() != 0.8 {
		t.Errorf("Diameter() = %v, want 0.8", tgt.Diameter())
	}
	if tgt.ScoringSystem() != targets.Custom {
		t.Errorf("ScoringSystem() = %q, want Custom", tgt.ScoringSystem())
	}
	got, _ := tgt.FaceSpec()
	want := targets.FaceSpec{0.1: 3, 0.5: 1}
	if !faceSpecEqual(got, want, 1e-9) {
		t.Errorf("FaceSpec() = %v, want %v", got, want)
	}
}

func TestCustomTargetFaceSpecUnits(t *testing.T) {
	spec := targets.FaceSpec{10: 5, 20: 4, 30: 3}
	tgt, err := targets.FromFaceSpec([2]any{spec, "cm"}, targets.CM(50), targets.Metres(30), false)
	if err != nil {
		t.Fatalf("FromFaceSpec error: %v", err)
	}
	got, _ := tgt.FaceSpec()
	want := targets.FaceSpec{0.1: 5, 0.2: 4, 0.3: 3}
	if !faceSpecEqual(got, want, 1e-9) {
		t.Errorf("FaceSpec() = %v, want %v", got, want)
	}
}

func TestCustomTargetInvalidFaceSpecUnits(t *testing.T) {
	spec := targets.FaceSpec{10: 5, 20: 4}
	_, err := targets.FromFaceSpec([2]any{spec, "bananas"}, targets.CM(50), targets.Metres(30), false)
	if err == nil {
		t.Fatal("expected error for unsupported face spec units")
	}
}

func TestCustomTargetEquality(t *testing.T) {
	base, _ := targets.FromFaceSpec(targets.FaceSpec{0.2: 2, 0.4: 1}, targets.CM(40), targets.Metres(20), true)
	dup, _ := targets.FromFaceSpec(targets.FaceSpec{0.2: 2, 0.4: 1}, targets.CM(40), targets.Metres(20), true)
	diff, _ := targets.FromFaceSpec(targets.FaceSpec{0.4: 5}, targets.CM(40), targets.Metres(20), true)

	if !base.Equal(dup) {
		t.Error("duplicate custom targets should be equal")
	}
	if base.Equal(diff) {
		t.Error("different custom targets should not be equal")
	}
}

func TestCustomTargetMaxScore(t *testing.T) {
	spec := targets.FaceSpec{0.02: 11, 0.04: 10, 0.8: 9, 0.12: 8, 0.16: 7, 0.2: 6}
	tgt, _ := targets.FromFaceSpec(spec, targets.CM(40), targets.Metres(18), false)
	if got := tgt.MaxScore(); got != 11 {
		t.Errorf("MaxScore() = %v, want 11", got)
	}
}

func TestCustomTargetMinScore(t *testing.T) {
	spec := targets.FaceSpec{0.02: 11, 0.04: 10, 0.8: 9, 0.12: 8, 0.16: 7, 0.2: 6}
	tgt, _ := targets.FromFaceSpec(spec, targets.CM(40), targets.Metres(18), false)
	if got := tgt.MinScore(); got != 6 {
		t.Errorf("MinScore() = %v, want 6", got)
	}
}

// ---- helpers ----

func approxEqual(a, b, tol float64) bool {
	d := a - b
	if d < 0 {
		d = -d
	}
	return d <= tol
}

func faceSpecEqual(a, b targets.FaceSpec, tol float64) bool {
	if len(a) != len(b) {
		return false
	}
	for k, va := range a {
		// Find closest key in b
		found := false
		for kb, vb := range b {
			if approxEqual(k, kb, tol) {
				if va != vb {
					return false
				}
				found = true
				break
			}
		}
		if !found {
			return false
		}
	}
	return true
}

File: ts/src/targets.test.ts
import { describe, it, expect } from 'vitest';
import { Target, ScoringSystem, genFaceSpec, cm, metres, yards, inches, type FaceSpec } from './targets.js';

// ---- helpers ----

function makeTarget(
  system: ScoringSystem,
  diamVal: number, diamUnit: string,
  distVal: number, distUnit: string,
  indoor = false,
): Target {
  return Target.create(system, { value: diamVal, units: diamUnit }, { value: distVal, units: distUnit }, indoor);
}

// ---- Target.toString ----

describe('Target.toString', () => {
  it('native units preserved in string', () => {
    const t = makeTarget(ScoringSystem.TenZone, 122, 'cm', 70, 'metre');
    expect(t.toString()).toContain('122');
    expect(t.toString()).toContain('cm');
    expect(t.toString()).toContain('70');
    expect(t.toString()).toContain('metre');
  });
  it('includes scoring system', () => {
    const t = makeTarget(ScoringSystem.FiveZone, 122, 'cm', 100, 'yard');
    expect(t.toString()).toContain('5_zone');
    expect(t.toString()).toContain('yard');
  });
});

// ---- Target equality ----

describe('Target equality', () => {
  it('identical targets are equal', () => {
    const a = makeTarget(ScoringSystem.TenZone, 122, 'cm', 70, 'metre');
    const b = makeTarget(ScoringSystem.TenZone, 122, 'cm', 70, 'metre');
    expect(a.equals(b)).toBe(true);
  });
  it('different scoring system', () => {
    const a = makeTarget(ScoringSystem.TenZone, 122, 'cm', 70, 'metre');
    const b = makeTarget(ScoringSystem.FiveZone, 122, 'cm', 70, 'metre');
    expect(a.equals(b)).toBe(false);
  });
  it('different distance', () => {
    const a = makeTarget(ScoringSystem.TenZone, 122, 'cm', 70, 'metre');
    const b = makeTarget(ScoringSystem.TenZone, 122, 'cm', 50, 'metre');
    expect(a.equals(b)).toBe(false);
  });
  it('different diameter', () => {
    const a = makeTarget(ScoringSystem.TenZone, 122, 'cm', 70, 'metre');
    const b = makeTarget(ScoringSystem.TenZone, 80, 'cm', 70, 'metre');
    expect(a.equals(b)).toBe(false);
  });
  it('different indoor flag', () => {
    const a = makeTarget(ScoringSystem.TenZone, 40, 'cm', 18, 'metre', false);
    const b = makeTarget(ScoringSystem.TenZone, 40, 'cm', 18, 'metre', true);
    expect(a.equals(b)).toBe(false);
  });
  it('yard and metre equal after conversion', () => {
    const a = makeTarget(ScoringSystem.FiveZone, 122, 'cm', 100, 'yard');
    const b = makeTarget(ScoringSystem.FiveZone, 122, 'cm', 91.44, 'metre');
    // different native units → not equal
    expect(a.equals(b)).toBe(false);
  });
});

// ---- Invalid construction ----

describe('Target.create validation', () => {
  it('throws for invalid scoring system', () => {
    expect(() => Target.create('unknown_system' as ScoringSystem, cm(122), metres(70), false))
      .toThrow('Invalid Target Face Type');
  });
  it('throws for invalid distance unit', () => {
    expect(() => Target.create(ScoringSystem.TenZone, cm(122), { value: 70, units: 'furlong' }, false))
      .toThrow();
  });
  it('throws for invalid diameter unit', () => {
    expect(() => Target.create(ScoringSystem.TenZone, { value: 122, units: 'furlong' }, metres(70), false))
      .toThrow();
  });
  it('defaults distance unit aliases', () => {
    const t = Target.create(ScoringSystem.TenZone, cm(122), { value: 70, units: 'metres' }, false);
    expect(t.nativeDistance.units).toBe('metre');
  });
});

// ---- Unit coercion to definitive ----

describe('unit coercion', () => {
  it('Yards → yard in native distance', () => {
    const t = makeTarget(ScoringSystem.FiveZone, 122, 'cm', 100, 'Yards');
    expect(t.nativeDistance.units).toBe('yard');
  });
  it('Metres → metre in native distance', () => {
    const t = makeTarget(ScoringSystem.TenZone, 122, 'cm', 70, 'Metres');
    expect(t.nativeDistance.units).toBe('metre');
  });
  it('CM → cm in native diameter', () => {
    const t = makeTarget(ScoringSystem.TenZone, 122, 'CM', 70, 'metre');
    expect(t.nativeDiameter.units).toBe('cm');
  });
});

// ---- Yard to metre conversion ----

describe('Target.distance in metres', () => {
  it('100 yards converts correctly', () => {
    const t = makeTarget(ScoringSystem.FiveZone, 122, 'cm', 100, 'yard');
    expect(t.distance).toBeCloseTo(91.44, 4);
  });
  it('metre distance unmodified', () => {
    const t = makeTarget(ScoringSystem.TenZone, 122, 'cm', 70, 'metre');
    expect(t.distance).toBe(70);
  });
});

// ---- Default location ----

describe('Target indoor flag', () => {
  it('defaults to outdoor (false)', () => {
    const t = Target.create(ScoringSystem.TenZone, cm(122), metres(70), false);
    expect(t.indoor).toBe(false);
  });
  it('indoor flag set', () => {
    const t = Target.create(ScoringSystem.Worcester, cm(40), metres(18), true);
    expect(t.indoor).toBe(true);
  });
});

// ---- maxScore for all systems ----

describe('Target.maxScore', () => {
  const cases: [ScoringSystem, number][] = [
    [ScoringSystem.FiveZone, 9],
    [ScoringSystem.TenZone, 10],
    [ScoringSystem.TenZoneCompound, 10],
    [ScoringSystem.TenZoneSixRing, 10],
    [ScoringSystem.TenZoneFiveRing, 10],
    [ScoringSystem.TenZoneFiveRingCmpd, 10],
    [ScoringSystem.ElevenZone, 11],
    [ScoringSystem.ElevenZoneSixRing, 11],
    [ScoringSystem.ElevenZoneFiveRing, 11],
    [ScoringSystem.WAField, 6],
    [ScoringSystem.IFAAField, 5],
    [ScoringSystem.IFAAFieldExpert, 5],
    [ScoringSystem.AANationalField, 5],
    [ScoringSystem.BeiterHitMiss, 1],
    [ScoringSystem.Worcester, 5],
    [ScoringSystem.Worcester2Ring, 5],
  ];
  for (const [sys, want] of cases) {
    it(`${sys} maxScore = ${want}`, () => {
      const t = Target.create(sys, cm(40), metres(18), false);
      expect(t.maxScore).toBe(want);
    });
  }
});

// ---- minScore ----

describe('Target.minScore', () => {
  it('5_zone minScore = 1', () => {
    expect(makeTarget(ScoringSystem.FiveZone, 122, 'cm', 70, 'metre').minScore).toBe(1);
  });
  it('10_zone minScore = 1', () => {
    expect(makeTarget(ScoringSystem.TenZone, 122, 'cm', 70, 'metre').minScore).toBe(1);
  });
  it('Worcester minScore = 1', () => {
    expect(makeTarget(ScoringSystem.Worcester, 40, 'cm', 18, 'metre', true).minScore).toBe(1);
  });
  it('Beiter_hit_miss minScore = 1', () => {
    expect(makeTarget(ScoringSystem.BeiterHitMiss, 40, 'cm', 18, 'metre').minScore).toBe(1);
  });
});

// ---- genFaceSpec / faceSpec ----

describe('genFaceSpec', () => {
  it('5_zone produces 5 rings with correct scores', () => {
    const spec = genFaceSpec(ScoringSystem.FiveZone, 1.22);
    expect(spec.size).toBe(5);
    const scores = [...spec.values()].sort((a, b) => a - b);
    expect(scores).toEqual([1, 3, 5, 7, 9]);
  });
  it('10_zone produces 10 rings', () => {
    const spec = genFaceSpec(ScoringSystem.TenZone, 1.22);
    expect(spec.size).toBe(10);
  });
  it('10_zone_6_ring produces 6 rings', () => {
    const spec = genFaceSpec(ScoringSystem.TenZoneSixRing, 1.22);
    expect(spec.size).toBe(6);
  });
  it('11_zone produces 11 rings', () => {
    const spec = genFaceSpec(ScoringSystem.ElevenZone, 0.40);
    expect(spec.size).toBe(11);
  });
  it('WA_field produces 6 rings', () => {
    const spec = genFaceSpec(ScoringSystem.WAField, 0.40);
    expect(spec.size).toBe(6);
  });
  it('Worcester produces 5 rings', () => {
    const spec = genFaceSpec(ScoringSystem.Worcester, 0.40);
    expect(spec.size).toBe(5);
  });
  it('Worcester_2_ring produces 2 rings', () => {
    const spec = genFaceSpec(ScoringSystem.Worcester2Ring, 0.40);
    expect(spec.size).toBe(2);
  });
  it('throws for unsupported system', () => {
    expect(() => genFaceSpec('unknown' as ScoringSystem, 1.0)).toThrow();
  });
});

// ---- Target.fromFaceSpec ----

describe('Target.fromFaceSpec', () => {
  it('custom face spec with cm units', () => {
    const spec: FaceSpec = new Map([[8, 10], [12, 8], [16, 7], [20, 6]]);
    const t = Target.fromFaceSpec([spec, 'cm'], cm(40), metres(18), true);
    expect(t.maxScore).toBe(10);
    expect(t.minScore).toBe(6);
    expect(t.indoor).toBe(true);
  });
  it('custom face spec in metres (default)', () => {
    const spec: FaceSpec = new Map([[0.08, 10], [0.12, 8], [0.16, 7], [0.20, 6]]);
    const t = Target.fromFaceSpec(spec, cm(40), metres(18), true);
    expect(t.maxScore).toBe(10);
    expect(t.faceSpec.size).toBe(4);
  });
  it('faceSpec copy is returned (not reference)', () => {
    const spec: FaceSpec = new Map([[0.08, 10], [0.12, 8]]);
    const t = Target.fromFaceSpec(spec, cm(40), metres(18), false);
    const returned = t.faceSpec;
    returned.set(99, 99);
    expect(t.faceSpec.has(99)).toBe(false);
  });
});

// ---- helper constructors ----

describe('Quantity helpers', () => {
  it('cm() creates cm quantity', () => {
    expect(cm(122)).toEqual({ value: 122, units: 'cm' });
  });
  it('metres() creates metre quantity', () => {
    expect(metres(70)).toEqual({ value: 70, units: 'metre' });
  });
  it('yards() creates yard quantity', () => {
    expect(yards(100)).toEqual({ value: 100, units: 'yard' });
  });
  it('inches() creates inch quantity', () => {
    expect(inches(16)).toEqual({ value: 16, units: 'inch' });
  });
});
